Add support for span name customization.

hyleung opened the pull request:
Add support for span name customization via module configuration. Module
configuration now can take a function that returns a custom span name
using information from the request and response.

Since the stock Ratpack response doesn't provide access to the
PathBinding (that's only accessible from the Context), we're using
wrapper to expose that information. If this ends up being supported
upstream, we can eventually get rid of this hack.

Closes #39 and #26

hyleung committed:
Add support for span name customization.

Add support for span name customization via module configuration. Module
configuration now can take a function that returns a custom span name
using information from the request and response.

Since the stock Ratpack response doesn't provide access to the
PathBinding (that's only accessible from the Context), we're using
wrapper to expose that information. If this ends up being supported
upstream, we can eventually get rid of this hack.

hyleung commented on src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java:
```diff
@@ -7,62 +7,103 @@
   }
 
-  static final class HttpAdapter extends brave.http.HttpServerAdapter<Request, Response> {
+  private static class ServerRequestImpl implements ServerRequest {
```
This diff is totally messed up, but the adapter has been moved out. Just felt a little tidier.

hyleung commented on src/main/java/ratpack/zipkin/ServerRequest.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Read-only API for wrappers around Ratpack's {@link ratpack.http.Request}.
+ */
+public interface ServerRequest {
```
Strictly speaking, this isn't necessary because Ratpack's `Request` gives us access to everything we would need. But I've kept this around because it felt like a read-only API in the lambda sort of communicates better what you can/cannot do (i.e don't try to mutate things in the lambda). Also for some reason it felt weird to have the lambda take a Ratpack `Request` for one param and a `ServerResponse` for the other.

hyleung commented on src/main/java/ratpack/zipkin/ServerTracingModule.java:
```diff
@@ -92,8 +96,10 @@ private static Endpoint buildLocalEndpoint(String serviceName, int port, @Nullab
+
     private HttpClientParser clientParser = new HttpClientParser();
     private HttpServerParser serverParser = new HttpServerParser();
+    private SpanNameProvider spanNameProvider = (req,res) -> req.getMethod().getName();
```
Default Span name is just the method name.

hyleung commented:
@llinder if you have any thoughts on this one, I'd greatly appreciate the feedback!

llinder commented:
Nice!  I read through this and nothing really suck out to comment on.  This should be useful for several folks that had some hacks on the old version to get access to the path binding.  Thanks for working on this!

codefromthecrypt commented on src/main/java/ratpack/zipkin/SpanNameProvider.java:
```diff
@@ -0,0 +1,15 @@
+   * @param response the response
+   * @return the Span name
+   */
+  String spanName(ServerRequest request, ServerResponse response);
```
response is odd to have here. seems the reason is to expose the path binding.. if so, make a comment or doc it so folks don't forget

hyleung replied:
@adriancole hmmm...yeah I see your point. TBH, there's probably not a hell of a lot of value in having `ServerResponse` here (I think I just have it this way because it seemed more "symmetrical"). I could just put the `PathBinding` in there - at least then it's pretty clear that that's the *only* reason that parameter exists (i.e. don't expect to be able to get anything else from the response).

codefromthecrypt commented:
The other thing is sometimes a failure happens before the response. Naming
based on the request makes sure the span is named even when a response
never comes.

On 17 Oct 2017 11:12, "Ho Yan Leung" <notifications@github.com> wrote:

> *@hyleung* commented on this pull request.
> ------------------------------
>
> In src/main/java/ratpack/zipkin/SpanNameProvider.java
> <https://github.com/hyleung/ratpack-zipkin/pull/42#discussion_r145018928>:
>
> > @@ -0,0 +1,15 @@
> +package ratpack.zipkin;
> +
> +/**
> + * Functional interface for span name customizers.
> + */
> +@FunctionalInterface
> +public interface SpanNameProvider {
> +  /**
> +   * Given a request and response, return the name to be used for the Span.
> +   * @param request the request
> +   * @param response the response
> +   * @return the Span name
> +   */
> +  String spanName(ServerRequest request, ServerResponse response);
>
> @adriancole <https://github.com/adriancole> hmmm...yeah I see your point.
> TBH, there's probably not a hell of a lot of value in having
> ServerResponse here (I think I just have it this way because it seemed
> more "symmetrical"). I could just put the PathBinding in there - at least
> then it's pretty clear that that's the *only* reason that parameter
> exists (i.e. don't expect to be able to get anything else from the
> response).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyleung/ratpack-zipkin/pull/42#discussion_r145018928>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD6182zVrbcmx_pfdstDx1fVxPir5UCks5stBsVgaJpZM4PdSi6>
> .
>

hyleung commented:
@adriancole ah, of course - I hadn't thought about that. That being the case, what do you think about making the second param `Optional<PathBinding>`?

llinder commented:
As @adriancole pointed out, applying the span name in the `beforeSend` handler is less than ideal.   I just don't see a way around this if we still want access to the `PathBinding`.  On the other hand, if `beforeSend` is never called its likely a framework bug so the chances of it not being called seems rather low.   Making it an `Optional<PathBinding>` does seem like a good idea though.  If I remember correctly, if Ratpack doesn't find a route that matches the request, the `PathBinding` might be null.   We should probably add some tests for this case if we don't have them yet.

If folks are concerned about naming the span in `beforeSend` it might be worth making the `SpanNameProvider` configurable so the `RatpackServerTracingHandler` can make a decision to sample in the request vs the response.

codefromthecrypt commented:
one thing I should mention is that it isn't bad to "rename" a span later,
just that it should have a name before the request goes out. For example,
nothing's stopping doing span.name("fooo") with a better name at response
time..  just if it has no name at all there could be more issues.

PS sampling needs to happen before the request goes out, as otherwise
downstream won't see the decision. (sampling is by default trace-scoped not
span-scoped though we can do span.abandon to throw away spans in a trace we
don't like)

On Tue, Oct 17, 2017 at 11:46 PM, Lance Linder <notifications@github.com>
wrote:

> As @adriancole <https://github.com/adriancole> pointed out, applying the
> span name in the beforeSend handler is less than ideal. I just don't see
> a way around this if we still want access to the PathBinding. On the
> other hand, if beforeSend is never called its likely a framework bug so
> the chances of it not being called seems rather low. Making it an
> Optional<PathBinding> does seem like a good idea though. If I remember
> correctly, if Ratpack doesn't find a route that matches the request, the
> PathBinding might be null. We should probably add some tests for this
> case if we don't have them yet.
>
> If folks are concerned about naming the span in beforeSend it might be
> worth making the SpanNameProvider configurable so the
> RatpackServerTracingHandler can make a decision to sample in the request
> vs the response.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hyleung/ratpack-zipkin/pull/42#issuecomment-337270755>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD6188Z4XIHRFmcN0wvBWOu3APNQc2Vks5stMu8gaJpZM4PdSi6>
> .
>

hyleung commented:
Ah...how about calling that span name provider (with `Optional<PathBinding>`) *twice* to set the span name - once when the request is received (with `Optional.empty()`) and once beforeSend (with a `PathBinding` value). That way we ensure that there'll be a span name regardless of whether there's a  failure.

The function that provides the span name would look something like this...

```
SpanNameProvider spanNameProvider = (req, pathBindingOpt) -> pathBindingOpt
    .map(pathBinding -> pathBinding.getDescription())
    .orElse(req.getPath());
```

hyleung committed:
Use Optonal PathBinding for span name customization

Instead of passing a wrapper around the Ratpack response to the
SpanNameCustomizer function, pass an Optional<PathBinding>.

Since the PathBinding is only available `beforeSend` on the RatPack response, we
call the customizer *twice* to ensure that the span will always have a name even
if for some reason an error occurs before the response is sent.

hyleung commented:
Alright, I reworked this to pass an `Optional<PathBinding>` //cc @llinder

codefromthecrypt commented:
For users trying to be efficient, they can check if the latter is present
and skip vs compute twice overriding. That or we can overload to do the
same. Or do nothing :)

hyleung committed:
Derp - forgot the first call to span name provider

:facepalm: (I'll squash this down later - just adding an extra commit for visibility)

llinder commented on src/main/java/ratpack/zipkin/internal/DefaultServerTracingHandler.java:
```diff
@@ -46,11 +46,11 @@ public DefaultServerTracingHandler(final HttpTracing httpTracing, final SpanName
     ctx.getResponse().beforeSend(response -> {
-      String name = spanNameProvider.spanName(request, Optional.of(ctx.getPathBinding()));
-      span.name(name);
+      span.name(spanNameProvider.spanName(request, Optional.of(ctx.getPathBinding())));
```
Not sure if `ctx.getPathBinding()` can ever be null but we can guard against an NPE here if we use `Optional.ofNullable`.

hyleung replied:
Btw, I couldn't figure out a way to force the failure case (where there's a failure before the response is sent) in a unit test ¯\_(ツ)_/¯

hyleung committed:
Use Optional.ofNullable in case PathBinding is null.

llinder commented:
Looks good!  Thanks again for the work on this.

hyleung committed:
Remove unused import

hyleung commented:
@llinder 👍 just did a quick (manual) test to make sure it's working as expected (https://github.com/hyleung/ratpack-zipkin-example/commit/0d426dcaaa34cfdba0b4502a5af10d1469457b2b). Did occur to me that #27 should probably provide access to the `PathBinding`, too. 

I'll let this sit for another day and then do a release and maybe start on that.
